Add lptmr driver support for NXP frdm_ke17z and frdm_ke17z512

ankeXiao opened the pull request:
Determine CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC default from devicetree.
Add support for KE1xZ SoC power management (idle, stop, partial stop 1, and partial stop 2), update the default system timer when PM is enabled.

Tested `drivers.counter.basic_api` and `tests/subsys/pm/power_mgmt_soc` sample.

ankeXiao committed:
dts: arm: nxp: nxp_ke1xz.dtsi: add lptmr support

Add lptmr support for ke17z, add related configuration for lptmr
driver.
Add supported CPU power states for idle, stop, partial stop 1, and
partial stop 2.

Signed-off-by: Anke Xiao <anke.xiao@nxp.com>

ankeXiao committed:
boards: nxp: add lptmr counter and timer features for KE17Z

Add LPTMR driver counter support for NXP frdm_ke17z and frdm_ke17z512
boards, tested 'drivers.counter.basic_api' case.
Enable cpu power states and determine the default
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC from devicetree when using the
Arm SysTick hardware timer.

Signed-off-by: Anke Xiao <anke.xiao@nxp.com>

decsny commented:
is the timer driver also going to be supported?

ankeXiao commented:
> is the timer driver also going to be supported?
Currently just to add lptmr driver support for these boards.
For timer feature, I am not sure, it may be supported in the future, I have not seen the sample of testing lptmr timer at present.

decsny commented:
> > is the timer driver also going to be supported?
> > Currently just to add lptmr driver support for these boards.
> > For timer feature, I am not sure, it may be supported in the future, I have not seen the sample of testing lptmr timer at present.

I believe LPTMR is the default system timer for the ke1xf soc series when power management is enabled. So every sample or test that enables  PM is testing LPTMR on that platform. And there is a timer api test specifically as well. In the past when we have enabled LPTMR on platforms, we make sure both the counter and timer driver work.

ankeXiao committed:
soc: nxp: kinetis: ke1xz: add support cpu power states

Enable the NXP Kinetis Low Power Timer (LPTMR) timer driver when
power management is enabled.
When the NXP KE1xZ SoC series is using the Arm SysTick as hardware
timer, the cycles/second will always be equal to the CPU core clock
frequency.

Signed-off-by: Anke Xiao <anke.xiao@nxp.com>

ankeXiao commented:
> > > is the timer driver also going to be supported?
> > > Currently just to add lptmr driver support for these boards.
> > > For timer feature, I am not sure, it may be supported in the future, I have not seen the sample of testing lptmr timer at present.
> 
> I believe LPTMR is the default system timer for the ke1xf soc series when power management is enabled. So every sample or test that enables PM is testing LPTMR on that platform. And there is a timer api test specifically as well. In the past when we have enabled LPTMR on platforms, we make sure both the counter and timer driver work.

Okay, I will refer to ke1xf to try to support lptmr timer driver, thanks!

ankeXiao committed:
tests: subsys: pm: add ke17z test support

Update power_mgmt_soc testcase.yaml to support test for frdm_ke17z
and frdm_ke17z512 platforms.

Signed-off-by: Anke Xiao <anke.xiao@nxp.com>

dleach02 commented on dts/arm/nxp/nxp_ke1xz.dtsi:
```diff
@@ -24,6 +24,32 @@
+				compatible = "zephyr,power-state";
+				power-state-name = "runtime-idle";
+			};
+
```
Do we not have any better state name then "suspend-to-idle" for each of the next three?

ankeXiao replied:
Align with ke1xf,rtxx, and rw6xx platforms, these states names are "suspend-to-idle", and looking at the power binding file, there seems to be no better name.

dleach02 commented on boards/nxp/frdm_ke17z/frdm_ke17z.dts:
```diff
@@ -60,6 +60,15 @@
+&stop {
+	min-residency-us = <20000>;
+	exit-latency-us = <13>;
+};
```
do we not need this information for pstop1 & pstop2?

ankeXiao replied:
No, partial Stop is a clocking option that can be taken instead of entering Stop mode and is configured in the SMC Stop Control Register (SMC_STOPCTRL).

ankeXiao commented:
Hi all, please help review this PR, we need merge this as soon as possible, thanks!

decsny commented on soc/nxp/kinetis/ke1xz/power.c:
```diff
@@ -0,0 +1,64 @@
+		/* Set partial stop mode and enable deep sleep */
+		SMC->STOPCTRL = SMC_STOPCTRL_PSTOPO(substate_id);
+		SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+#ifdef CONFIG_XIP
```
is there a reason why you can't use if(IS_ENABLED()) and allow optimization instead of preprocessor directives (preprocessor directives tend to lead to configuration dependency problems and confusing issues in the long term evolution of C code, need to avoid when possible, even if it seems simple now)

ankeXiao replied:
Thanks for the reminder, this looks better, I'll replace it.

ankeXiao commented:
@decsny @dleach02  , I have re-updated the PR, please help review. Thanks!

ankeXiao commented:
Can someone help merge this PR? Thanks!

ankeXiao commented:
Hello everyone, please help to review this pull request so that we can merge it as soon as possible. The merger of PR requires at least two approved reviews, thanks!
